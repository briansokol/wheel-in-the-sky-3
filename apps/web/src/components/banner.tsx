import { TinyColor } from '@ctrl/tinycolor';
import { Segment } from '@repo/shared/classes/segment';
import { WheelManager } from '@repo/shared/classes/wheel-manager';
import { RefObject, useEffect, useMemo } from 'react';

interface BannerProps {
    bannerRef: RefObject<HTMLCanvasElement | null>;
    wheelManager: WheelManager | undefined;
    winner: Segment | undefined;
}

const FONT_NAME = '"Quicksand", sans-serif';

/**
 * Banner component that displays the wheel title and winner in a styled canvas
 * with rounded corners and triangular indicators pointing inward
 */
export function Banner({ bannerRef, wheelManager, winner }: BannerProps) {
    const fgColor = useMemo(() => {
        return winner?.textColor ?? '#ffffff';
    }, [winner?.textColor]);

    const segmentColor = useMemo(() => {
        return winner?.backgroundColor ?? '#000000';
    }, [winner?.backgroundColor]);

    const bgColor = useMemo(() => {
        if (!winner) {
            return '#000000';
        }

        const color = new TinyColor(winner.backgroundColor);
        const brightness = color.getBrightness();
        const isLight = color.isLight();

        if (isLight) {
            if (brightness > 215) {
                return color.darken(20).toString();
            } else {
                return color.brighten(20).toString();
            }
        } else {
            if (brightness < 30) {
                return color.brighten(20).toString();
            } else {
                return color.darken(20).toString();
            }
        }
    }, [winner]);

    useEffect(() => {
        if (bannerRef.current) {
            const ctx = bannerRef.current.getContext('2d');
            if (ctx) {
                const canvas = bannerRef.current;
                const canvasWidth = canvas.width;
                const canvasHeight = canvas.height;
                const cornerRadius = 15; // Radius for rounded corners

                // Clear the canvas to start with transparency
                ctx.clearRect(0, 0, canvasWidth, canvasHeight);

                // Create a rounded rectangle path for the background
                ctx.beginPath();
                ctx.moveTo(cornerRadius, 0);
                ctx.lineTo(canvasWidth - cornerRadius, 0);
                ctx.arcTo(canvasWidth, 0, canvasWidth, cornerRadius, cornerRadius);
                ctx.lineTo(canvasWidth, canvasHeight - cornerRadius);
                ctx.arcTo(canvasWidth, canvasHeight, canvasWidth - cornerRadius, canvasHeight, cornerRadius);
                ctx.lineTo(cornerRadius, canvasHeight);
                ctx.arcTo(0, canvasHeight, 0, canvasHeight - cornerRadius, cornerRadius);
                ctx.lineTo(0, cornerRadius);
                ctx.arcTo(0, 0, cornerRadius, 0, cornerRadius);
                ctx.closePath();

                // Fill the rounded rectangle background
                ctx.fillStyle = bgColor;
                ctx.fill();

                // Draw isosceles triangles pointing toward center
                const triangleWidth = 350;
                const triangleHeight = 250;
                const verticalPosition = canvasHeight / 2 - triangleHeight / 2;

                // Left triangle (pointing right)
                ctx.beginPath();
                ctx.fillStyle = segmentColor;
                ctx.moveTo(0, verticalPosition);
                ctx.lineTo(0, verticalPosition + triangleHeight);
                ctx.lineTo(triangleWidth, verticalPosition + triangleHeight / 2);
                ctx.closePath();
                ctx.fill();

                // Right triangle (pointing left)
                ctx.beginPath();
                ctx.fillStyle = segmentColor;
                ctx.moveTo(canvasWidth, verticalPosition);
                ctx.lineTo(canvasWidth, verticalPosition + triangleHeight);
                ctx.lineTo(canvasWidth - triangleWidth, verticalPosition + triangleHeight / 2);
                ctx.closePath();
                ctx.fill();

                // Set text properties for text
                ctx.fillStyle = fgColor;
                ctx.textAlign = 'center';
                ctx.textBaseline = 'middle';

                // Draw "Winner!" text
                ctx.font = `28px ${FONT_NAME}`;
                ctx.fillText('Winner!', canvasWidth / 2, canvasHeight / 3);

                // Draw title text
                ctx.fillText(wheelManager?.config.fullTitle ?? '', canvasWidth / 2, (canvasHeight * 2) / 3);

                // Draw winner name
                ctx.font = `48px ${FONT_NAME}`;
                ctx.fillText(winner?.name ?? '', canvasWidth / 2, canvasHeight / 2);

                // Draw app name
                ctx.font = `16px ${FONT_NAME}`;
                ctx.fillText('wheelinthesky.app', canvasWidth / 2, canvasHeight - 25);
            }
        }
    }, [bannerRef, bgColor, fgColor, segmentColor, wheelManager, winner]);

    return (
        <canvas
            ref={bannerRef}
            width="800"
            height="300"
            style={{
                borderRadius: '15px',
            }}
        ></canvas>
    );
}
