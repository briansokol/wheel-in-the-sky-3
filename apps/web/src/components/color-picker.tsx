import { Button, Input } from '@heroui/react';
import type { FormColorNameValue } from '@repo/shared/types/config';
import { debounce } from 'es-toolkit';
import { type FC, useCallback, useMemo } from 'react';
import { HexColorPicker } from 'react-colorful';
import { useFormContext } from 'react-hook-form';

interface ColorPickerProps {
    singleColorName: FormColorNameValue;
    multiColorName?: FormColorNameValue;
    allowMultiColor: boolean;
}

export const ColorPicker: FC<ColorPickerProps> = ({ singleColorName, multiColorName = '', allowMultiColor }) => {
    const { setValue, watch } = useFormContext();

    const debouncedSetColor = useMemo(() => {
        return debounce((value: string) => setValue(singleColorName, value), 100);
    }, [singleColorName, setValue]);

    const color: string = watch(singleColorName);
    const watchColors = watch(multiColorName);
    const colors: string[] = useMemo(
        () => (multiColorName !== undefined && watchColors !== undefined ? JSON.parse(watchColors) : []),
        [multiColorName, watchColors]
    );

    const addColor = useCallback(() => {
        if (multiColorName !== undefined && !colors.includes(color)) {
            const newColors = [...colors, color];
            setValue(multiColorName, JSON.stringify(newColors));
        }
    }, [color, colors, multiColorName, setValue]);

    const removeColor = useCallback(
        (index: number) => () => {
            if (multiColorName !== undefined) {
                const newColors = [...colors];
                newColors.splice(index, 1);
                setValue(multiColorName, JSON.stringify(newColors));
            }
        },
        [colors, multiColorName, setValue]
    );

    return (
        <>
            <HexColorPicker color={color} onChange={debouncedSetColor} style={{ width: '100%' }} />
            <div className="my-3">
                <Input
                    className="mr-3 inline-block w-[120px]"
                    type="text"
                    value={color}
                    onChange={(e) => setValue(singleColorName, e.target.value)}
                    aria-label="Chosen Color"
                />
                {allowMultiColor && multiColorName && (
                    <Button className="mr-3" onPress={addColor} color="default">
                        Add Color
                    </Button>
                )}
                {colors && colors.length > 0 && 'Click a color to remove it'}
            </div>
            {allowMultiColor && multiColorName && (
                <div data-testid="picker-color-list" className="flex flex-wrap">
                    {colors.map((color, i) => (
                        <Button
                            key={color}
                            className="relative mb-3 mr-3"
                            style={{ background: color }}
                            onPress={removeColor(i)}
                        ></Button>
                    ))}
                </div>
            )}
        </>
    );
};
