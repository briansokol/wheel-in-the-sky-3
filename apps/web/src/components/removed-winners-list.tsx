import { Button } from '@heroui/react';
import { useCallback, useContext } from 'react';
import { MdRestore } from 'react-icons/md';
import { RemovedWinnersContext } from '@/contexts/removed-winners';

/**
 * Component that displays a list of removed winners with the ability to restore them.
 */
export function RemovedWinnersList() {
    const { removedWinners, setRemovedWinners } = useContext(RemovedWinnersContext);

    /**
     * Restores a winner by removing them from the removed winners list.
     *
     * @param {string} winner - The winner to restore
     * @returns {void}
     */
    const restoreWinner = useCallback(
        (winner: string) => {
            setRemovedWinners?.((prevWinners) => prevWinners.filter((w) => w !== winner));
        },
        [setRemovedWinners]
    );

    return (
        <div className="flex flex-col gap-2">
            {removedWinners.map((winner) => (
                <div key={winner} className="flex items-center justify-between">
                    <p className="max-w-[80%] truncate" title={winner}>
                        {winner}
                    </p>
                    <Button
                        isIconOnly
                        onPress={() => {
                            restoreWinner(winner);
                        }}
                    >
                        <MdRestore />
                    </Button>
                </div>
            ))}
            {removedWinners.length === 0 && (
                <div>
                    <p className="text-center text-gray-500">No removed winners</p>
                </div>
            )}
            <div>
                <Button
                    className="w-full"
                    isDisabled={removedWinners.length === 0}
                    startContent={<MdRestore />}
                    onPress={() => {
                        setRemovedWinners?.([]);
                    }}
                >
                    Restore All
                </Button>
            </div>
        </div>
    );
}
