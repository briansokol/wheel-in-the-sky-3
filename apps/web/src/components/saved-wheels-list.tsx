import { Button, ButtonGroup } from '@heroui/react';
import { useContext, useMemo } from 'react';
import { FaCircle } from 'react-icons/fa6';
import { MdSave } from 'react-icons/md';
import { MdDeleteForever } from 'react-icons/md';
import { MdOutlinePlayCircleFilled } from 'react-icons/md';
import { useNavigate } from 'react-router';
import { ConfigContext } from '@/contexts/config';
import { SegmentContext } from '@/contexts/segment';
import { dateIsWithinLast5Minutes } from '@/utils/dates';
import { useSavedWheels } from '@/utils/local-storage';

interface SavedWheelsListProps {
    closeDrawer?: () => void;
}

export function SavedWheelsList({ closeDrawer }: SavedWheelsListProps) {
    const { encodedConfig, decodedConfig } = useContext(ConfigContext);
    const { setHasWinner } = useContext(SegmentContext);
    const { savedWheels, saveWheel, removeSavedWheel } = useSavedWheels();
    const navigate = useNavigate();

    const currentWheelIsSaved = useMemo(() => {
        return savedWheels.some((wheel) => wheel.id === decodedConfig?.id);
    }, [savedWheels, decodedConfig]);

    return (
        <div className="flex flex-col gap-4">
            {savedWheels.map((wheel) => (
                <div key={wheel.id} className="flex w-full items-center gap-2">
                    <div className="shrink-0">
                        <div
                            className={`rounded-full border-2 p-1 ${
                                wheel.createdAt === wheel.updatedAt &&
                                dateIsWithinLast5Minutes(new Date(wheel.createdAt))
                                    ? 'border-green-500'
                                    : dateIsWithinLast5Minutes(new Date(wheel.updatedAt))
                                      ? 'border-blue-500'
                                      : 'border-transparent'
                            } `}
                            title={`${wheel.id === decodedConfig?.id ? 'Active Wheel' : 'Saved Wheel'}${
                                wheel.createdAt === wheel.updatedAt &&
                                dateIsWithinLast5Minutes(new Date(wheel.createdAt))
                                    ? ' (Recently Saved)'
                                    : dateIsWithinLast5Minutes(new Date(wheel.updatedAt))
                                      ? ' (Recently Updated)'
                                      : ''
                            }`}
                        >
                            <FaCircle
                                className={`text-2xl ${wheel.id === decodedConfig?.id ? 'text-primary' : 'text-gray-700'}`}
                            />
                        </div>
                    </div>
                    <div className="min-w-0 flex-1 overflow-hidden">
                        <p className="truncate">{wheel.title}</p>
                        {wheel.description && (
                            <p className="truncate text-sm italic text-gray-400">{wheel.description}</p>
                        )}
                    </div>
                    <div>
                        <ButtonGroup>
                            <Button
                                isIconOnly
                                variant="flat"
                                color="primary"
                                onPress={() => {
                                    setHasWinner?.(false);
                                    navigate(`/wheel/v3/${encodeURIComponent(wheel.encodedConfig)}`);
                                    closeDrawer?.();
                                }}
                                className="text-2xl"
                                title="Load this Wheel"
                            >
                                <MdOutlinePlayCircleFilled />
                            </Button>
                            <Button
                                isIconOnly
                                variant="flat"
                                color="danger"
                                onPress={() => {
                                    removeSavedWheel(wheel.id);
                                }}
                                className="text-2xl"
                                title="Delete Saved Wheel"
                            >
                                <MdDeleteForever />
                            </Button>
                        </ButtonGroup>
                    </div>
                </div>
            ))}
            <Button
                className="w-full text-base"
                isDisabled={currentWheelIsSaved || !decodedConfig || !encodedConfig}
                startContent={<MdSave className="text-2xl" />}
                color="primary"
                onPress={() => {
                    saveWheel(decodedConfig, encodedConfig);
                }}
            >
                Save Current Wheel
            </Button>
            <div>
                <p className="text-center text-sm text-gray-500">
                    Wheels are saved locally to your browser. They are not synced between devices.
                </p>
            </div>
        </div>
    );
}
