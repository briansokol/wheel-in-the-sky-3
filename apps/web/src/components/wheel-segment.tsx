import { Segment } from '@repo/shared/classes/segment';
import { type CSSProperties, type FC, memo, useMemo } from 'react';

interface WheelSegmentProps {
    segment: Segment;
    showName?: boolean;
}

export const WheelSegment: FC<WheelSegmentProps> = ({ segment, showName = true }) => {
    const styles = useMemo(() => {
        const baseStyles: CSSProperties = {
            backgroundColor: segment.backgroundColor,
            transform: `rotate(-90deg) rotate(${segment.degrees / 2}deg) rotate(${segment.degrees * segment.index}deg)`,
            transformOrigin: '50% 50%',
        };

        if (segment.degrees === 180) {
            baseStyles.clipPath = `polygon(100% 0, 100% 100%, 50% 100%, 50% 0)`;
        } else if (segment.degrees < 180) {
            baseStyles.clipPath = `polygon(50% 50%, 100% ${segment.legStart}%, 100% ${segment.legEnd}%)`;
        }

        return baseStyles;
    }, [segment]);

    return (
        <div
            data-testid="wheel-segment"
            className="absolute top-0 box-border size-full text-right font-semibold text-black"
            style={styles}
        >
            <div
                data-testid="wheel-segment-name"
                className="absolute right-5 top-1/2 -translate-y-1/2 text-right text-base [transform-origin:right_center] sm:text-xl"
                style={{ color: segment.textColor }}
            >
                {showName ? segment.name : ''}
            </div>
        </div>
    );
};

export const MemoizedWheelSegment = memo(WheelSegment);
