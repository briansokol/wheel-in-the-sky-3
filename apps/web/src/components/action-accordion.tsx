import { Accordion, AccordionItem } from '@heroui/react';
import { PageBaseRoute } from '@/constants/routes';
import { isPage } from '@/utils/routes';
import { RemovedWinnersList } from './removed-winners-list';
import { SavedWheelsList } from './saved-wheels-list';

interface ActionAccordionProps {
    pathname: string;
    closeDrawer?: () => void;
}

export function ActionAccordion({ pathname, closeDrawer }: ActionAccordionProps) {
    return (
        <Accordion selectionMode="multiple" defaultExpandedKeys={['saved-wheels', 'removed-winners']}>
            <AccordionItem key="saved-wheels" aria-label="Saved Wheels" title="Saved Wheels">
                <SavedWheelsList closeDrawer={closeDrawer} />
            </AccordionItem>
            <AccordionItem key="removed-winners" aria-label="Removed Winners" title="Removed Winners">
                {isPage(pathname, PageBaseRoute.WheelV3) ? <RemovedWinnersList closeDrawer={closeDrawer} /> : null}
            </AccordionItem>
        </Accordion>
    );
}
