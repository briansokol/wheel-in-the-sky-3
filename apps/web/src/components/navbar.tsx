import {
    Button,
    Divider,
    Drawer,
    DrawerBody,
    DrawerContent,
    DrawerHeader,
    Link,
    Navbar,
    NavbarBrand,
    NavbarContent,
    NavbarItem,
    NavbarMenu,
    NavbarMenuItem,
    NavbarMenuToggle,
    useDisclosure,
} from '@heroui/react';
import { useCallback, useMemo, useState } from 'react';
import { FaList } from 'react-icons/fa6';
import { useLocation, useParams } from 'react-router';
import { PageBaseRoute } from '@/constants/routes';
import { isPage } from '@/utils/routes';
import { ActionAccordion } from './action-accordion';

export function AppNavBar() {
    const { pathname } = useLocation();
    const params = useParams();
    const configId = params?.id ?? 'new';
    const uriEncodedConfigId = useMemo(() => encodeURIComponent(configId), [configId]);

    const [isMenuOpen, setIsMenuOpen] = useState(false);
    const { isOpen: isDrawerOpen, onOpen: onDrawerOpen, onOpenChange: onDrawerOpenChange } = useDisclosure();

    const closeMenu = useCallback(() => {
        setIsMenuOpen(false);
    }, [setIsMenuOpen]);

    const buildNavbarItems = useCallback(
        (Component: typeof NavbarItem | typeof NavbarMenuItem) => {
            return (
                <>
                    {configId !== 'new' && (
                        <Component isActive={isPage(pathname, PageBaseRoute.WheelV3)}>
                            <Link
                                color="foreground"
                                aria-current={isPage(pathname, PageBaseRoute.WheelV3) ? 'page' : 'false'}
                                href={`/wheel/v3/${uriEncodedConfigId}`}
                                onPress={closeMenu}
                            >
                                Wheel
                            </Link>
                        </Component>
                    )}
                    <Component isActive={isPage(pathname, PageBaseRoute.ConfigV3)}>
                        <Link
                            color="foreground"
                            aria-current={isPage(pathname, PageBaseRoute.ConfigV3) ? 'page' : 'false'}
                            href={`/config/v3/${uriEncodedConfigId}`}
                            onPress={closeMenu}
                        >
                            {configId === 'new' ? 'Create Wheel' : 'Change Wheel'}
                        </Link>
                    </Component>
                    <Component isActive={isPage(pathname, PageBaseRoute.About)}>
                        <Link
                            color="foreground"
                            aria-current={isPage(pathname, PageBaseRoute.About) ? 'page' : 'false'}
                            href={`/about`}
                            onPress={closeMenu}
                        >
                            About
                        </Link>
                    </Component>
                </>
            );
        },
        [configId, uriEncodedConfigId, pathname, closeMenu]
    );

    const navbarItems = useMemo(() => {
        return buildNavbarItems(NavbarItem);
    }, [buildNavbarItems]);

    const navbarMenuItems = useMemo(() => {
        return buildNavbarItems(NavbarMenuItem);
    }, [buildNavbarItems]);

    return (
        <>
            <Navbar
                isMenuOpen={isMenuOpen}
                onMenuOpenChange={setIsMenuOpen}
                classNames={{
                    item: [
                        'flex',
                        'relative',
                        'h-full',
                        'items-center',
                        "data-[active=true]:after:content-['']",
                        'data-[active=true]:after:absolute',
                        'data-[active=true]:after:bottom-0',
                        'data-[active=true]:after:left-0',
                        'data-[active=true]:after:right-0',
                        'data-[active=true]:after:h-[2px]',
                        'data-[active=true]:after:rounded-[2px]',
                        'data-[active=true]:after:bg-primary',
                    ],
                }}
            >
                <NavbarMenuToggle aria-label={isMenuOpen ? 'Close Menu' : 'Open Menu'} className="sm:hidden" />
                <NavbarBrand>
                    <Link
                        color="foreground"
                        aria-current={isPage(pathname, PageBaseRoute.Home) ? 'page' : 'false'}
                        href={`/`}
                    >
                        <p className="font-bold text-inherit">Wheel in the Sky</p>
                    </Link>
                </NavbarBrand>
                <NavbarContent className="hidden gap-4 sm:flex" justify="center">
                    {navbarItems}
                </NavbarContent>
                <NavbarContent className="hidden gap-4 sm:flex" justify="end">
                    <NavbarItem>
                        <Button variant="light" endContent={<FaList />} onPress={onDrawerOpen}>
                            Actions
                        </Button>
                    </NavbarItem>
                </NavbarContent>
                <NavbarMenu>
                    {navbarMenuItems}
                    <Divider />
                    <ActionAccordion pathname={pathname} closeDrawer={closeMenu} />
                </NavbarMenu>
            </Navbar>
            <Drawer isOpen={isDrawerOpen} onOpenChange={onDrawerOpenChange} size="md">
                <DrawerContent>
                    {() => (
                        <>
                            <DrawerHeader className="flex flex-col gap-1">Actions</DrawerHeader>
                            <DrawerBody>
                                <ActionAccordion pathname={pathname} closeDrawer={onDrawerOpenChange} />
                            </DrawerBody>
                        </>
                    )}
                </DrawerContent>
            </Drawer>
        </>
    );
}
