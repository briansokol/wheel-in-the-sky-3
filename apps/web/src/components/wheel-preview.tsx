import { TinyColor } from '@ctrl/tinycolor';
import { Card } from '@heroui/react';
import { Config } from '@repo/shared/classes/config';
import { WheelManager } from '@repo/shared/classes/wheel-manager';
import { DEFAULT_FOREGROUND_COLOR, INVERSE_FOREGROUND_COLOR } from '@repo/shared/constants/colors';
import { PageColorType } from '@repo/shared/enums/page-colors';
import { WheelColorType } from '@repo/shared/enums/wheel-colors';
import { buildPageColorConfig, buildWheelColorConfig } from '@repo/shared/utils/configs';
import { FC, useContext, useEffect, useMemo } from 'react';
import { Wheel } from '@/components/wheel';
import { RotationContext } from '@/contexts/rotation';
import { useBgColor, useFgColor } from '@/hooks/colors';
import { calculateStartingRotation } from '@/utils/math';

interface WheelPreviewProps {
    title: string;
    description: string;
    names: string;
    randomizeOrder: boolean;
    wheelColorType: WheelColorType;
    baseColor?: string;
    colors?: string[];
    randomizeColor?: boolean;
    pageColorType: PageColorType;
    backgroundColor: string;
    showNames: boolean;
    isSelected?: boolean;
    height?: string;
}

export const WheelPreview: FC<WheelPreviewProps> = ({
    title,
    names,
    randomizeOrder,
    wheelColorType,
    baseColor,
    colors,
    randomizeColor,
    pageColorType,
    backgroundColor,
    showNames,
    height,
}) => {
    const { setRotation } = useContext(RotationContext);

    const [config, wheelManager] = useMemo(() => {
        const newConfig = new Config();
        newConfig.wheelColorConfig = buildWheelColorConfig(
            wheelColorType,
            baseColor ?? '',
            colors ?? [],
            randomizeColor ?? false
        );

        let foregroundColor = '';
        if (pageColorType === PageColorType.Single) {
            foregroundColor = new TinyColor(backgroundColor).isDark()
                ? DEFAULT_FOREGROUND_COLOR
                : INVERSE_FOREGROUND_COLOR;
        }

        newConfig.pageColorConfig = buildPageColorConfig(pageColorType, backgroundColor, foregroundColor);

        newConfig.setNames(names);
        newConfig.setRandomizeOrder(randomizeOrder);
        newConfig.setShowNames(showNames);

        const newWheelManager = new WheelManager();
        newWheelManager.init(newConfig, []);

        return [newConfig, newWheelManager];
    }, [
        names,
        randomizeOrder,
        wheelColorType,
        baseColor,
        colors,
        randomizeColor,
        pageColorType,
        backgroundColor,
        showNames,
    ]);

    useEffect(() => {
        if (setRotation && wheelManager) {
            setRotation(calculateStartingRotation(wheelManager?.segments?.length ?? 0));
        }
    }, [setRotation, wheelManager]);

    const bgColor = useBgColor(config, config.pageColorConfig.backgroundColor);
    const fgColor = useFgColor(config, config.pageColorConfig.foregroundColor);

    return (
        <Card className="w-full" shadow="lg" style={{ height: height ?? '400px', background: bgColor }}>
            {names.length > 0 ? (
                <div className="flex h-full flex-col items-center justify-center overflow-hidden">
                    <div
                        className="flex translate-y-8 items-center justify-center text-xl transition-all"
                        style={{ color: fgColor }}
                    >
                        {title}
                    </div>
                    <div className="flex size-full translate-y-28 items-center justify-center transition-all">
                        <Wheel wheelManager={wheelManager} radius="70%" isStatic={true} />
                    </div>
                </div>
            ) : (
                <div className="flex h-full items-center justify-center overflow-hidden" style={{ color: fgColor }}>
                    Add at Least 1 Wheel Segment
                </div>
            )}
        </Card>
    );
};
