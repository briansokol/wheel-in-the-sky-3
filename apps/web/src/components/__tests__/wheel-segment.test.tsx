import { TinyColor } from '@ctrl/tinycolor';
import { Segment } from '@repo/shared/classes/segment';
import { render, screen } from '@testing-library/react';
import { WheelSegment } from '@/components/wheel-segment';

describe('WheelSegment', () => {
    let mockSegment: Segment;
    beforeEach(() => {
        mockSegment = new Segment(0, 'Test Segment', 90, 0, 90, new TinyColor('#FF0000'));
    });

    it('should render segment with correct name', () => {
        render(<WheelSegment segment={mockSegment} />);
        expect(screen.getByText('Test Segment')).toBeInTheDocument();
    });

    it('should apply correct base styles', () => {
        const { container } = render(<WheelSegment segment={mockSegment} />);
        const segmentDiv = container.firstChild as HTMLElement;

        expect(segmentDiv).toHaveStyle({
            backgroundColor: '#ff0000',
            transform: 'rotate(-90deg) rotate(45deg) rotate(0deg)',
            transformOrigin: '50% 50%',
        });
    });

    it('should not apply clip path when it is the only segment', () => {
        mockSegment = new Segment(0, 'Test Segment', 360, 0, 90, new TinyColor('#FF0000'));

        const { container } = render(<WheelSegment segment={mockSegment} />);
        const segmentDiv = container.firstChild as HTMLElement;

        expect(segmentDiv).toHaveStyle({clipPath:''});
    });

    it('should apply proper clip path when there are only 2 segments', () => {
        mockSegment = new Segment(0, 'Test Segment', 180, 0, 90, new TinyColor('#FF0000'));

        const { container } = render(<WheelSegment segment={mockSegment} />);
        const segmentDiv = container.firstChild as HTMLElement;

        expect(segmentDiv.style.clipPath).toContain('polygon(100% 0');
    });

    it('should apply proper clip path when there are more than 3 segments', () => {
        const { container } = render(<WheelSegment segment={mockSegment} />);
        const segmentDiv = container.firstChild as HTMLElement;

        expect(segmentDiv.style.clipPath).toContain('polygon(50% 50%');
    });

    it('should render text with correct styling', () => {
        const { container } = render(<WheelSegment segment={mockSegment} />);
        const textDiv = container.querySelector('.absolute.right-5') as HTMLElement;

        expect(textDiv).toHaveStyle({
            color: mockSegment.textColor,
            transformOrigin: 'right center',
            transform: 'translateY(-50%)',
        });
    });

    it('should handle different segment indices', () => {
        const mockSegment2 = new Segment(2, 'Test Segment 2', 90, 180, 270, new TinyColor('#00FF00'));

        const { container } = render(<WheelSegment segment={mockSegment2} />);
        const segmentDiv = container.firstChild as HTMLElement;

        expect(segmentDiv).toHaveStyle({
            transform: 'rotate(-90deg) rotate(45deg) rotate(180deg)',
        });
    });

    it('should handle different segment sizes', () => {
        const mockSegment3 = new Segment(0, 'Test Segment 3', 60, 0, 60, new TinyColor('#0000FF'));

        const { container } = render(<WheelSegment segment={mockSegment3} />);
        const segmentDiv = container.firstChild as HTMLElement;

        expect(segmentDiv).toHaveStyle({
            transform: 'rotate(-90deg) rotate(30deg) rotate(0deg)',
        });
    });

    it('matches snapshots', () => {
        // 1 segment
        const mockSegment1 = new Segment(0, 'Test Segment 1', 360, 0, 0, new TinyColor('#0000FF'));
        const { container: container1 } = render(<WheelSegment segment={mockSegment1} />);
        expect(container1.firstChild).toMatchSnapshot();

        // 2 segments
        const mockSegment2 = new Segment(0, 'Test Segment 2', 360 / 2, 0, 0, new TinyColor('#0000FF'));
        const { container: container2 } = render(<WheelSegment segment={mockSegment2} />);
        expect(container2.firstChild).toMatchSnapshot();

        // 3 segments
        const mockSegment3 = new Segment(0, 'Test Segment 3', 360 / 3, 0, 0, new TinyColor('#0000FF'));
        const { container: container3 } = render(<WheelSegment segment={mockSegment3} />);
        expect(container3.firstChild).toMatchSnapshot();

        // 4 segments
        const mockSegment4 = new Segment(0, 'Test Segment 4', 360 / 4, 0, 0, new TinyColor('#0000FF'));
        const { container: container4 } = render(<WheelSegment segment={mockSegment4} />);
        expect(container4.firstChild).toMatchSnapshot();

        // 5 segments
        const mockSegment5 = new Segment(0, 'Test Segment 5', 360 / 5, 0, 0, new TinyColor('#0000FF'));
        const { container: container5 } = render(<WheelSegment segment={mockSegment5} />);
        expect(container5.firstChild).toMatchSnapshot();
    });
});
