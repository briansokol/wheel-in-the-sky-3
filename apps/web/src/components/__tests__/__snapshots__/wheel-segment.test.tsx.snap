// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WheelSegment > matches snapshots 1`] = `
<div
  class="absolute top-0 box-border size-full text-right font-semibold text-black"
  data-testid="wheel-segment"
  style="background-color: rgb(0, 0, 255); transform: rotate(-90deg) rotate(180deg) rotate(0deg); transform-origin: 50% 50%;"
>
  <div
    class="absolute right-5 top-1/2 -translate-y-1/2 text-right text-base [transform-origin:right_center] sm:text-xl"
    data-testid="wheel-segment-name"
    style="color: rgb(255, 255, 255);"
  >
    Test Segment 1
  </div>
</div>
`;

exports[`WheelSegment > matches snapshots 2`] = `
<div
  class="absolute top-0 box-border size-full text-right font-semibold text-black"
  data-testid="wheel-segment"
  style="background-color: rgb(0, 0, 255); transform: rotate(-90deg) rotate(90deg) rotate(0deg); transform-origin: 50% 50%; clip-path: polygon(100% 0, 100% 100%, 50% 100%, 50% 0);"
>
  <div
    class="absolute right-5 top-1/2 -translate-y-1/2 text-right text-base [transform-origin:right_center] sm:text-xl"
    data-testid="wheel-segment-name"
    style="color: rgb(255, 255, 255);"
  >
    Test Segment 2
  </div>
</div>
`;

exports[`WheelSegment > matches snapshots 3`] = `
<div
  class="absolute top-0 box-border size-full text-right font-semibold text-black"
  data-testid="wheel-segment"
  style="background-color: rgb(0, 0, 255); transform: rotate(-90deg) rotate(60deg) rotate(0deg); transform-origin: 50% 50%; clip-path: polygon(50% 50%, 100% -36.602540000000005%, 100% 136.60254%);"
>
  <div
    class="absolute right-5 top-1/2 -translate-y-1/2 text-right text-base [transform-origin:right_center] sm:text-xl"
    data-testid="wheel-segment-name"
    style="color: rgb(255, 255, 255);"
  >
    Test Segment 3
  </div>
</div>
`;

exports[`WheelSegment > matches snapshots 4`] = `
<div
  class="absolute top-0 box-border size-full text-right font-semibold text-black"
  data-testid="wheel-segment"
  style="background-color: rgb(0, 0, 255); transform: rotate(-90deg) rotate(45deg) rotate(0deg); transform-origin: 50% 50%; clip-path: polygon(50% 50%, 100% 0%, 100% 100%);"
>
  <div
    class="absolute right-5 top-1/2 -translate-y-1/2 text-right text-base [transform-origin:right_center] sm:text-xl"
    data-testid="wheel-segment-name"
    style="color: rgb(255, 255, 255);"
  >
    Test Segment 4
  </div>
</div>
`;

exports[`WheelSegment > matches snapshots 5`] = `
<div
  class="absolute top-0 box-border size-full text-right font-semibold text-black"
  data-testid="wheel-segment"
  style="background-color: rgb(0, 0, 255); transform: rotate(-90deg) rotate(36deg) rotate(0deg); transform-origin: 50% 50%; clip-path: polygon(50% 50%, 100% 13.672870000000003%, 100% 86.32713%);"
>
  <div
    class="absolute right-5 top-1/2 -translate-y-1/2 text-right text-base [transform-origin:right_center] sm:text-xl"
    data-testid="wheel-segment-name"
    style="color: rgb(255, 255, 255);"
  >
    Test Segment 5
  </div>
</div>
`;
